api/konnect/v1alpha1/konnect_cloudgateway_transitgateway_types.go

```diff
@@ -0,0 +1,168 @@
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:resource:categories=kong
+// +kubebuilder:object:root=true
+// +kubebuilder:object:generate=true
+// +kubebuilder:subresource:status
+// +kubebuilder:subresource:finalizer
+// +apireference:kgo:include
+// +kong:channels=gateway-operator
+type KonnectCloudGatewayTransitGateway struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of KonnectCloudGatewayTransitGateway.
+	Spec KonnectCloudGatewayTransitGatewaySpec `json:"spec"`
+
+	// Status defines the observed state of KonnectCloudGatewayTransitGateway.
+	Status KonnectCloudGatewayTransitGatewayStatus `json:"status,omitempty"`
+}
+
+// KonnectCloudGatewayTransitGatewaySpec defines the desired state of KonnectCloudGatewayTransitGateway.
+//
+// +kubebuilder:validation:XValidation:rule="self.type == oldSelf.type",message="spec.type is immutable"
+// +kubebuilder:validation:XValidation:rule="self.type == 'AWS Transit Gateway' ? self.attachment_config.kind == 'aws-transit-gateway-attachment' : true",message="AWS transit gateway must have spec.attachment_config.kind set to 'aws-transit-gateway-attachment'"
+// +kubebuilder:validation:XValidation:rule="self.type == 'AWS AWS VPC Peering Gateway' ? self.attachment_config.kind == 'aws-vpc-peering-attachment' : true",message="AWS VPC peering gateway must have spec.attachment_config.kind set to 'aws-vpc-peering-attachment'"
+// +kubebuilder:validation:XValidation:rule="self.type == 'Azure Transit Gateway' ? self.attachment_config.kind == 'azure-vnet-peering-attachment' : true",message="Azure VNET peering gateway must have spec.attachment_config.kind set to 'azure-vnet-peering-attachment'"
+// TODO: add more constraints on attachment_config based on type
+type KonnectCloudGatewayTransitGatewaySpec struct {
+	// NetworkRef is the schema for the NetworkRef type.
+	//
+	// +kubebuilder:validation:Required
+	NetworkRef commonv1alpha1.ObjectRef `json:"networkRef"`
+
+	KonnectTransitGatewayAPISpec `json:",inline"`
+}
+
+// KonnectTransitGatewayAPISpec specifies a transit gateway on the Konnect side.
+// The type and all the types it referenced are mostly copied github.com/Kong/sdk-konnect-go/models/components.CreateTransitGatewayRequest.
+// We consolidated the common fields other than 'attachment_config' because k8s CRDs does not accept slices with the same json keys in different types.
+//
+// TODO: add necessary comments for types/fields if they were missing in sdk-konnect-go.
+type KonnectTransitGatewayAPISpec struct {
+	// Type is the type of the Konnect transit gateway.
+	// Currently types AWS Transit Gateway, AWS VPC Peering Gateway, Azure Transit Gateway are supported.
+	//
+	// +kubebuilder:validation:Enum=AWS Transit Gateway;AWS VPC Peering Gateway;Azure Transit Gateway
+	Type sdkkonnectcomp.CreateTransitGatewayRequestType `json:"type"`
+	// Human-readable name of the transit gateway.
+	Name string `json:"name"`
```

Comment:
As per https://github.com/JoelSpeed/kal?tab=readme-ov-file#optionalorrequired, all fields should be either optional or required. 

Let's make sure to add `// +kubebuilder:validation:Required` or `// +kubebuilder:validation:Optional` to each field to adhere to the linked ruleset.